core/io/json.cpp

```diff
@@ -453,5 +448,18 @@ Error JSON::parse(const String &p_json, Variant &r_ret, String &r_err_str, int &
 
 	err = _parse_value(r_ret, token, str, idx, len, r_err_line, r_err_str);
 
+	// Check if EOF is reached
+	// or it's a type of the next token.
+	if (err == OK && idx < len) {
+		err = _get_token(str, idx, len, token, r_err_line, r_err_str);
+
+		if (err || token.type != TK_EOF) {
+			r_err_str = "Expected 'EOF'";
+			// Reset return value to empty `Variant`
+			r_ret = Variant();
+			return ERR_PARSE_ERROR;
```

Comment:
Yeah just tested now, I confirm it resets the value to `Variant()` (aka `null` in GDScript), all assertions in #40795 pass!

This is the case where I'm not really sure whether returning a "half-way" token even on parse error is intended, but I'm not sure if this "bug" was used as a "feature" before. 🙂